### PR TITLE
Refactor common logic to filters in solution API controller

### PR DESCRIPTION
**Pull request by ErikSchierboom**
I found some duplication in the solutions API controller whilst working on the unlock_help button, and thought I'd dry it up a bit.

- Refactor use_solution logic into before_action
- Refactor use_user_track logic into before_action

### CODE REVIEW

File: app/controllers/api/solutions_controller.rb [hunk 1]
```diff
@@ -163,37 +108,37 @@ def diff
...
-    def set_track
-      @track = Track.find(params[:track_slug])
+      return render_solution_not_accessible unless @solution.user_id == current_user.id
     end
 
-    def set_exercise
-      @exercise = @track.exercises.find(params[:exercise_slug])
+    def use_user_track
+      @user_track = UserTrack.for(current_user, @solution.track)
+      return render_404(:track_not_joined) if @user_track.external?
```
Comment (iHiD) [inline app/controllers/api/solutions_controller.rb]:
```suggestion
      render_404(:track_not_joined) if @user_track.external?
```

File: app/controllers/api/solutions_controller.rb [hunk 2]
```diff
@@ -163,37 +108,37 @@ def diff
...
-
-      Solution::UpdateToLatestExerciseVersion.(solution)
 
-      render json: { solution: SerializeSolution.(solution) }
-    end
-
-    private
-    def set_track
-      @track = Track.find(params[:track_slug])
+      return render_solution_not_accessible unless @solution.user_id == current_user.id
```
Comment (iHiD) [inline app/controllers/api/solutions_controller.rb]:
```suggestion
      render_solution_not_accessible unless @solution.user_id == current_user.id
```

File: app/controllers/api/solutions_controller.rb [hunk 3]
```diff
@@ -163,37 +108,37 @@ def diff
...
+      render json: { solution: SerializeSolution.(@solution) }
+    end
+
+    private
+    def use_solution
       begin
-        solution = Solution.find_by!(uuid: params[:uuid])
+        @solution = Solution.find_by!(uuid: params[:uuid])
       rescue ActiveRecord::RecordNotFound
         return render_solution_not_found
```
Comment (iHiD) [inline app/controllers/api/solutions_controller.rb:133]:
```suggestion
        render_solution_not_found
```

**Review by iHiD — APPROVED** [review]
Two small tweaks.

**Comment by ErikSchierboom** [comment]
I had to adjust the suggestion a tiny bit (move the accessible code within the begin/end block). I also removed another used method

**Pull request merged**